File: src/Sample/Plumix.Sample/Demos/General/StatefulBuilderLookupBoundaryDemoPage.cs
using Avalonia;
using Avalonia.Media;
using Plumix.Foundation;
using Plumix.Material;
using Plumix.Rendering;
using Plumix.Widgets;

namespace Plumix;

// Dart parity source: dart_sample/lib/demos/general/stateful_builder_lookup_boundary_demo_page.dart

public sealed class StatefulBuilderLookupBoundaryDemoPage : StatelessWidget
{
    public override Widget Build(BuildContext context)
    {
        int count = 0;
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 16,
            children:
            [
                new Text("StatefulBuilder + LookupBoundary", fontSize: 20, color: Colors.Black),
                new Text(
                    "StatefulBuilder owns a local rebuild. LookupBoundary hides ancestors only from its bounded " +
                    "static lookup helpers.",
                    fontSize: 14,
                    color: Colors.DimGray),
                new Container(
                    color: Color.Parse("#FFF4F7FA"),
                    padding: new Thickness(12),
                    child: new StatefulBuilder((builderContext, setState) =>
                        new Column(
                            crossAxisAlignment: CrossAxisAlignment.Start,
                            spacing: 8,
                            children:
                            [
                                new Text($"local count: {count}", color: Color.Parse("#FF31506F")),
                                new TextButton(
                                    onPressed: () => setState(() => count += 1),
                                    child: new Text("Increment local state")),
                            ]))),
                new DemoLookupScope(
                    label: "outer scope",
                    child: new LookupBoundary(
                        child: new Builder(builderContext =>
                        {
                            string bounded = LookupBoundary
                                .GetInheritedWidgetOfExactType<DemoLookupScope>(builderContext)
                                ?.Label ?? "hidden";
                            string regular = builderContext.DependOnInherited<DemoLookupScope>()?.Label ?? "missing";
                            return new Container(
                                color: Color.Parse("#FFE7EDF6"),
                                padding: new Thickness(12),
                                child: new Column(
                                    crossAxisAlignment: CrossAxisAlignment.Start,
                                    spacing: 6,
                                    children:
                                    [
                                        new Text($"bounded lookup: {bounded}", color: Colors.Black),
                                        new Text($"regular context lookup: {regular}", color: Colors.Black),
                                    ]));
                        }))),
            ]);
    }

    private sealed class DemoLookupScope : InheritedWidget
    {
        public DemoLookupScope(string label, Widget child, Key? key = null) : base(key)
        {
            Label = label;
            Child = child;
        }

        public string Label { get; }

        public Widget Child { get; }

        public override Widget Build(BuildContext context) => Child;

        protected override bool UpdateShouldNotify(InheritedWidget oldWidget)
        {
            return Label != ((DemoLookupScope)oldWidget).Label;
        }
    }
}

File: dart_sample/lib/demos/general/stateful_builder_lookup_boundary_demo_page.dart
import 'package:material_ui/material_ui.dart';

class StatefulBuilderLookupBoundaryDemoPage extends StatelessWidget {
  const StatefulBuilderLookupBoundaryDemoPage({super.key});

  @override
  Widget build(BuildContext context) {
    int count = 0;
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 16,
      children: <Widget>[
        const Text(
          'StatefulBuilder + LookupBoundary',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'StatefulBuilder owns a local rebuild. LookupBoundary hides ancestors '
          'only from its bounded static lookup helpers.',
          style: TextStyle(fontSize: 14, color: Colors.black54),
        ),
        Container(
          color: const Color(0xFFF4F7FA),
          padding: const EdgeInsets.all(12),
          child: StatefulBuilder(
            builder: (BuildContext context, StateSetter setState) {
              return Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                spacing: 8,
                children: <Widget>[
                  Text(
                    'local count: $count',
                    style: const TextStyle(color: Color(0xFF31506F)),
                  ),
                  TextButton(
                    onPressed: () => setState(() => count += 1),
                    child: const Text('Increment local state'),
                  ),
                ],
              );
            },
          ),
        ),
        _DemoLookupScope(
          label: 'outer scope',
          child: LookupBoundary(
            child: Builder(
              builder: (BuildContext context) {
                final String bounded =
                    LookupBoundary.getInheritedWidgetOfExactType<
                          _DemoLookupScope
                        >(context)
                        ?.label ??
                    'hidden';
                final String regular =
                    context
                        .dependOnInheritedWidgetOfExactType<_DemoLookupScope>()
                        ?.label ??
                    'missing';
                return Container(
                  color: const Color(0xFFE7EDF6),
                  padding: const EdgeInsets.all(12),
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    spacing: 6,
                    children: <Widget>[
                      Text('bounded lookup: $bounded'),
                      Text('regular context lookup: $regular'),
                    ],
                  ),
                );
              },
            ),
          ),
        ),
      ],
    );
  }
}

class _DemoLookupScope extends InheritedWidget {
  const _DemoLookupScope({required this.label, required super.child});

  final String label;

  @override
  bool updateShouldNotify(_DemoLookupScope oldWidget) {
    return label != oldWidget.label;
  }
}
